CSS-4381 Adding the openfga interface definition. (#78)
canonical/charm-relation-interfaces: interfaces/openfga/v0/README.md

# `openfga`

## Usage

This relation interface describes the expected behaviour of any charm claiming to be able to interact with a OpenFGA.

> In most cases, this will be accomplished using the [openfga library](https://github.com/canonical/cs-openfga/blob/main/charms/openfga-k8s/lib/charms/openfga_k8s/v0/openfga.py), although charm developers are free to provide alternative libraries as long as they fulfil the behavioural and schematic requirements described in this document.

Reviewer: Is there a reason for the linked repo to be private?

Reviewer 2: no, it should be public as agreed a few weeks ago

Author: it was made public last week

## Direction

```mermaid
flowchart TD

>     Requirer -- openfga, store-name --> Provider

Reviewer 3: what is this `openfga` field being passed around? Didn't see it in the databags.

    Provider -- openfga, store-id, token, address, scheme, port --> Requirer
```

> As with all Juju relations, the `openfga` interface consists of two parties: a Provider (openfga charm), and a Requirer (application charm). The Requirer will be expected to provide an authentication store name, and the Provider will provide new unique credentials (along with other optional fields), which can be used to access the OpenFGA store.

Reviewer 3: ```suggestion
As with all Juju relations, the `openfga` interface consists of two parties: a Provider (openfga charm), and a Requirer (application charm). The Requirer will be expected to expose an authentication store name, and the Provider will create and forward new unique credentials (along with other optional fields), which can be used to access the OpenFGA store.
```

tip: avoid using `provide` and `require` in these documents, they're kind of confusing in the context :)

## Behavior

> Both the Requirer and the Provider need to adhere to criteria to be considered compatible with the interface.

Reviewer 3: ```suggestion
Both the Requirer and the Provider need to adhere to the following criteria to be considered compatible with the interface.
```

### Provider
- Is expected to create an authentication store in OpenFGA when the requirer provides the `store_name` field.
- Is expected to expose to the Requirer `token_secret_id`, `store_id`, `address`, `scheme` and `port` fields in the *application* databag.

### Requirer

> 

Reviewer: ```suggestion
- Is expected to provide the `store_name` it requires in its application databag
```

- Is expected to provide an authentication store name in the `store_name` field of the *application* databag.

## Relation Data

### Provider

[\[JSON Schema\]](./schemas/provider.json)

Provider provides `token`, `store_id`, `address`, `scheme` and `port` fields. It should be placed in the **application** databag.

#### Example
```yaml
  relation-info:
  - endpoint: openfga
    related-endpoint: openfga
    application-data:

>       token: test-token

Author: but that means that the charm is limited to juju versions that implement secrets.

>       token: test-token

Reviewer: Correct (Juju > 3.0). You intend this new interface to be compatible with Juju 2.9?

>       token: test-token

Author: i suggest we keep v0 as is and move v1 to using secrets once we have a prodstack running juju 3.0+

>       token: test-token

Reviewer: If we know for sure that this interface will be used by Juju<3.0, I agree, we don't really have a choice. However if we're not sure, I'd recommend going directly with secrets. From your last comments, it looks like it needs to support Juju 2.9

>       token: test-token

Author: done

      store_id: 01GK13VYZK62Q1T0X55Q2BHYD6
      address: "10.10.0.17"
      scheme: http

>       port: "8080"

Reviewer: Please standardise aroung the quotation marks. Ex. scheme and store_id don't have them and address and port do have them.

Author: done

```

### Requirer

[\[JSON Schema\]](./schemas/requirer.json)

> Requirer provides authorization store name. It should be placed in the **application** databag.

Reviewer 3: ```suggestion
The Requirer exposes the store name for which authorization is requested in the **application** databag.
```
same as above (please double-check phrasing, not sure I understand what openfga does)
